### DISCUSSION TITLE
Cannot register custom exception handler

### DISCUSSION THREAD

**Discussion by lgedgar** [flet-dev/flet#5735] [Q&A]
### Duplicate Check

- [X] I have searched the [opened issues](https://github.com/flet-dev/flet/issues) and there are no duplicates

### Describe the bug

I need to register a "catch-all" exception handler, to help identify bugs in development.  Some errors may possibly arise from Flet controls but more often they will be result of backend API calls failing etc.

I previously did this via `sys.excepthook` and it did the trick fine, but this was under Flet 0.19.0 and it appears this approach no longer works with Flet 0.21.2+

I then turned to setting `Page.on_error` but this either does not work, or at least does not do what I was hoping.

I of course prefer to use latest Flet but currently this feels less "safe" given I have no insight to new errors.

### Code sample

<details open><summary>Logs</summary>

```python
# -*- coding: utf-8; -*-

import sys
import threading

import flet as ft


def sample_main(page):

    def handle_error(exc_type, exc_value, exc_traceback):
        print("\n\nGOT UNHANDLED ERROR\n\n")

    def sys_exc_hook(exc_type, exc_value, exc_traceback):
        handle_error(exc_type, exc_value, exc_traceback)
        sys.__excepthook__(exc_type, exc_value, exc_traceback)

    sys.excepthook = sys_exc_hook

    # custom exception hook for threads (requires python 3.8)
    # cf. https://docs.python.org/3/library/threading.html#threading.excepthook
    v = sys.version_info
    if v.major >= 3 and v.minor >= 8:

        def thread_exc_hook(args):
            handle_error(args.exc_type, args.exc_value, args.exc_traceback)
            # nb. as of python 3.10 could just call threading.__excepthook__ instead
            # cf. https://docs.python.org/3/library/threading.html#threading.__excepthook__
            orig_thread_hook(args)

        # nb. as of python 3.10 the original hook is accessible, we needn't save the ref
        # cf. https://docs.python.org/3/library/threading.html#threading.__excepthook__
        orig_thread_hook = threading.excepthook

        threading.excepthook = thread_exc_hook

    def page_error(e):
        print('\n\nGOT PAGE ERROR:', e, '\n')

    page.on_error = page_error

    def on_click(e):
        print('GOT BUTTON CLICK', e)
        raise RuntimeError("whoops")

    page.add(ft.ElevatedButton("click me", on_click=on_click))


def run_sample_app():
    ft.app(target=sample_main)


if __name__ == '__main__':
    run_sample_app()
```

</details>

### To reproduce

Running the code above with Flet 0.19.0 *does* print "GOT UNHANDLED ERROR" when clicking the button.  But running it with newer Flet, the button click error still raises and a traceback is printed to console, but no "GOT UNHANDLED ERROR" message and so my custom hook is being ignored somehow.

I did try registering just the sys hook and then just the threading hook, to see which was actually getting called under Flet 0.19.0 and it seems the threading hook was being called, not the sys hook.  But again neither gets called with newer Flet.

I then added the `Page.on_error` snippet in the hopes it would get called, but it also is not getting called - this is true with both the older and newer Flet.

### Expected behavior

I need a reliable way to detect unhandled errors, so I can "handle" them in a minimal way.  Details of the error will be logged but also sent via email, and I need to show a brief message to the user indicating that an error was encountered.  All of this did work with Flet 0.19.0 and a custom `excepthook` but now I'm not sure how to register a handler that actually works?

Docs for [Page.on_error](https://flet.dev/docs/controls/page/#on_error) makes me think this is what I should perhaps be using instead of hacking global `excepthook`, but https://github.com/flet-dev/flet/issues/431 makes me think this was always meant as a more "Flet-specific" error handler and perhaps *not* what I would want?

I also looked over https://docs.flutter.dev/testing/errors but do not yet understand enough about the Flutter side to fully grok what this means for me in Flet-land.

### Screenshots / Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Operating System

Linux

### Operating system details

Debian 12

### Flet version

0.23.2

### Regression

Yes, it used to work in a previous Flet version (please specify the version in additional details)

### Suggestions

I'm not 100% sure that my desired approach is the correct one; for instance `Page.on_error` seems promising instead of the `excepthook` overrides.  And/or maybe a global error handler is not the best way to solve this in the first place?  But it has worked okay in the past with other (non-Flet) projects..

Ideally:

- i discover how to correctly register "global" error handler
- possibly, Flet docs are updated accordingly
- possibly, Flet code must change to accommodate (e.g. `Page.on_error` improved) ?

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Additional details

Global threading `excepthook` used to work under Flet 0.19.0

Also - the above is all using Python 3.11.2

**Comment by nickstuer** [comment]
I know this is old but in case someone else comes across this, here is a work around:

``` 
import pprint
import os
import traceback

def exception_handler(loop, context):
  print("error")
  pprint.pprint(context['exception'])
  logger.info(f"{context['exception']}: {''.join(traceback.format_tb(context['exception'].__traceback__))}")
  self.page.window_destroy()

async def sample_main(page: ft.Page):
  loop = asyncio.get_running_loop()
  loop.set_exception_handler(exception_handler)```

Must add async keyword.
